client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -226,20 +229,25 @@ boolean isWithinRefreshThreshold(Instant currentInstant, Instant expiration) {
         long traceEnterId = LoggerHelpers.traceEnter(log, "refreshToken", this.scopeName, this.streamName);
         CompletableFuture<Void> currentRefreshFuture = tokenRefreshFuture.get();
         if (currentRefreshFuture == null) {
-            log.debug("Initiated token refresh for scope {} and stream {}", this.scopeName, this.streamName);
+            log.debug("Initiating token refresh for scope {} and stream {}", this.scopeName, this.streamName);
             currentRefreshFuture = this.recreateToken();
             this.tokenRefreshFuture.compareAndSet(null, currentRefreshFuture);
         } else {
             log.debug("Token is already under refresh for scope {} and stream {}", this.scopeName, this.streamName);
         }
 
         final CompletableFuture<Void> handleToCurrentRefreshFuture  = currentRefreshFuture;
-        return currentRefreshFuture.thenApply(v -> {
-                    // Token is already refreshed, so resetting the future to null.
-                    this.tokenRefreshFuture.compareAndSet(handleToCurrentRefreshFuture, null);
-                    LoggerHelpers.traceLeave(log, "refreshToken", traceEnterId, this.scopeName, this.streamName);
-                    return delegationToken.get().getValue();
-                });
+        return currentRefreshFuture.handle((v, ex) -> {
+            this.tokenRefreshFuture.compareAndSet(handleToCurrentRefreshFuture, null);
+            LoggerHelpers.traceLeave(log, "refreshToken", traceEnterId, this.scopeName, this.streamName);
+            if (ex != null) {
+                log.warn("Encountered an exception in when refreshing token for scope {} and stream {}",
+                        this.scopeName, this.streamName, Exceptions.unwrap(ex));
+                throw ex instanceof CompletionException ? (CompletionException) ex : new CompletionException(ex);
```

Comment:
we can simply throw `ex` wrapped inside a CompletationException.

Reply:
The intention of throwing the existing exception is to avoid creating a new exception object unnecessarily.

Reply:
if you dont want to "handle" the returned value/exception, then you can use `whenComplete`

Reply:
ok. `controllerClient.getOrRefreshDelegationTokenFor(scopeName, streamName)` will not throw an exception of type CompletationException.